File: app/routers/rent_agreements.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List

from app.database import get_db
from app import crud, schemas, auth, models

router = APIRouter()

@router.post("/", response_model=schemas.RentAgreementResponse)
def create_rent_agreement(
    agreement: schemas.RentAgreementCreate,
    db: Session = Depends(get_db),
    current_user: schemas.UserResponse = Depends(auth.get_current_active_user)
):
    """Create a new rent agreement"""
    # Check if property exists
    property = crud.get_property(db, agreement.property_id)
    if not property:
        raise HTTPException(status_code=404, detail="Property not found")
    
    # Check if property is already occupied
    if property.is_occupied:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Property is already occupied"
        )
    
    # Check if there's already an active agreement for this property
    existing_agreement = crud.get_active_rent_agreement_by_property(db, agreement.property_id)
    if existing_agreement:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Property already has an active rent agreement"
        )
    
    # Only property owner can create agreements
    if property.owner_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only property owner can create rent agreements"
        )
    
    # Check if tenant exists
    tenant = crud.get_user(db, agreement.tenant_id)
    if not tenant:
        raise HTTPException(status_code=404, detail="Tenant not found")
    
    if tenant.user_type != models.UserType.TENANT:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Selected user is not a tenant"
        )
    
    return crud.create_rent_agreement(db=db, agreement=agreement)

@router.get("/my-agreements", response_model=List[schemas.RentAgreementResponse])
def get_my_agreements(
    skip: int = 0,
    limit: int = 100,
    db: Session = Depends(get_db),
    current_user: schemas.UserResponse = Depends(auth.get_current_active_user)
):
    """Get rent agreements for current user (as tenant or as property owner)"""
    if current_user.user_type == models.UserType.TENANT:
        return crud.get_rent_agreements_by_tenant(db=db, tenant_id=current_user.id, skip=skip, limit=limit)
    else:
        # For owners, get agreements for their properties
        agreements = []
        properties = crud.get_properties_by_owner(db=db, owner_id=current_user.id, skip=0, limit=1000)
        for property in properties:
            prop_agreements = crud.get_rent_agreements_by_property(db=db, property_id=property.id, skip=0, limit=100)
            agreements.extend(prop_agreements)
        return agreements[skip:skip+limit]

@router.get("/property/{property_id}", response_model=List[schemas.RentAgreementResponse])
def get_agreements_by_property(
    property_id: int,
    skip: int = 0,
    limit: int = 100,
    db: Session = Depends(get_db),
    current_user: schemas.UserResponse = Depends(auth.get_current_active_user)
):
    """Get rent agreements for a specific property"""
    # Check if property exists
    property = crud.get_property(db, property_id)
    if not property:
        raise HTTPException(status_code=404, detail="Property not found")
    
    # Only property owner can view agreements
    if property.owner_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Not enough permissions"
        )
    
    return crud.get_rent_agreements_by_property(db=db, property_id=property_id, skip=skip, limit=limit)

@router.get("/{agreement_id}", response_model=schemas.RentAgreementResponse)
def get_rent_agreement(
    agreement_id: int,
    db: Session = Depends(get_db),
    current_user: schemas.UserResponse = Depends(auth.get_current_active_user)
):
    """Get rent agreement by ID"""
    agreement = crud.get_rent_agreement(db, agreement_id=agreement_id)
    if agreement is None:
        raise HTTPException(status_code=404, detail="Rent agreement not found")
    
    # Check permissions - either tenant or property owner can view
    property = crud.get_property(db, agreement.property_id)
    if current_user.id != agreement.tenant_id and current_user.id != property.owner_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Not enough permissions"
        )
    
    return agreement

@router.get("/", response_model=List[schemas.RentAgreementResponse])
def get_rent_agreements(
    skip: int = 0,
    limit: int = 100,
    db: Session = Depends(get_db),
    current_user: schemas.UserResponse = Depends(auth.get_current_active_user)
):
    """Get list of all rent agreements (admin only for now)"""
    agreements = crud.get_rent_agreements(db, skip=skip, limit=limit)
    return agreements

@router.put("/{agreement_id}", response_model=schemas.RentAgreementResponse)
def update_rent_agreement(
    agreement_id: int,
    agreement_update: schemas.RentAgreementUpdate,
    db: Session = Depends(get_db),
    current_user: schemas.UserResponse = Depends(auth.get_current_active_user)
):
    """Update rent agreement"""
    # Check if agreement exists
    agreement = crud.get_rent_agreement(db, agreement_id=agreement_id)
    if agreement is None:
        raise HTTPException(status_code=404, detail="Rent agreement not found")
    
    # Check permissions - only property owner can update
    property = crud.get_property(db, agreement.property_id)
    if current_user.id != property.owner_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only property owner can update rent agreements"
        )
    
    updated_agreement = crud.update_rent_agreement(db, agreement_id=agreement_id, agreement_update=agreement_update)
    return updated_agreement

@router.post("/{agreement_id}/terminate")
def terminate_rent_agreement(
    agreement_id: int,
    termination_reason: str,
    db: Session = Depends(get_db),
    current_user: schemas.UserResponse = Depends(auth.get_current_active_user)
):
    """Terminate rent agreement"""
    # Check if agreement exists
    agreement = crud.get_rent_agreement(db, agreement_id=agreement_id)
    if agreement is None:
        raise HTTPException(status_code=404, detail="Rent agreement not found")
    
    # Check permissions - either tenant or property owner can terminate
    property = crud.get_property(db, agreement.property_id)
    if current_user.id != agreement.tenant_id and current_user.id != property.owner_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Not enough permissions"
        )
    
    terminated_agreement = crud.terminate_rent_agreement(db, agreement_id=agreement_id, termination_reason=termination_reason)
    if terminated_agreement is None:
        raise HTTPException(status_code=404, detail="Rent agreement not found")
    
    return {"message": "Rent agreement terminated successfully"}

File: src/routes/rent-agreements.ts
import { Router, Request, Response } from 'express';
import { db } from '@/services/database';
import { requireAuth, requireActiveUser, getCurrentUser } from '@/middleware/auth';
import { validate, schemas } from '@/utils/validation';
import { ApiResponse, RentAgreement, RentAgreementCreate } from '@/types';
import { logger } from '@/utils/logger';

const router = Router();

/**
 * POST /api/v1/rent-agreements
 * Create a new rent agreement
 */
router.post('/', requireActiveUser, validate(schemas.rentAgreementCreate), async (req: Request, res: Response) => {
  try {
    const currentUser = getCurrentUser(req);
    
    if (!currentUser) {
      const response: ApiResponse = {
        success: false,
        message: 'User not found',
      };
      return res.status(404).json(response);
    }

    const agreementData: RentAgreementCreate = req.body;
    const agreement = await db.rentAgreement.create(agreementData);

    const response: ApiResponse<RentAgreement> = {
      success: true,
      data: agreement,
      message: 'Rent agreement created successfully',
    };

    res.status(201).json(response);
  } catch (error) {
    logger.error('Create rent agreement failed:', error);
    
    const response: ApiResponse = {
      success: false,
      message: 'Failed to create rent agreement',
      error: error instanceof Error ? error.message : 'Unknown error',
    };

    res.status(500).json(response);
  }
});

/**
 * GET /api/v1/rent-agreements
 * Get all rent agreements for current user
 */
router.get('/', requireActiveUser, async (req: Request, res: Response) => {
  try {
    const currentUser = getCurrentUser(req);
    
    if (!currentUser) {
      const response: ApiResponse = {
        success: false,
        message: 'User not found',
      };
      return res.status(404).json(response);
    }

    const agreements = await db.rentAgreement.findByTenantId(currentUser.id);

    const response: ApiResponse<RentAgreement[]> = {
      success: true,
      data: agreements,
    };

    res.status(200).json(response);
  } catch (error) {
    logger.error('Get rent agreements failed:', error);
    
    const response: ApiResponse = {
      success: false,
      message: 'Failed to get rent agreements',
      error: error instanceof Error ? error.message : 'Unknown error',
    };

    res.status(500).json(response);
  }
});

/**
 * GET /api/v1/rent-agreements/:id
 * Get rent agreement by ID
 */
router.get('/:id', requireActiveUser, async (req: Request, res: Response) => {
  try {
    const agreementId = parseInt(req.params.id);
    
    if (isNaN(agreementId)) {
      const response: ApiResponse = {
        success: false,
        message: 'Invalid agreement ID',
      };
      return res.status(400).json(response);
    }

    const currentUser = getCurrentUser(req);
    
    if (!currentUser) {
      const response: ApiResponse = {
        success: false,
        message: 'User not found',
      };
      return res.status(404).json(response);
    }

    const agreement = await db.rentAgreement.findById(agreementId);
    
    if (!agreement) {
      const response: ApiResponse = {
        success: false,
        message: 'Rent agreement not found',
      };
      return res.status(404).json(response);
    }

    // Check if user is authorized to view this agreement
    if (agreement.tenantId !== currentUser.id) {
      const response: ApiResponse = {
        success: false,
        message: 'Unauthorized to view this agreement',
      };
      return res.status(403).json(response);
    }

    const response: ApiResponse<RentAgreement> = {
      success: true,
      data: agreement,
    };

    res.status(200).json(response);
  } catch (error) {
    logger.error('Get rent agreement failed:', error);
    
    const response: ApiResponse = {
      success: false,
      message: 'Failed to get rent agreement',
      error: error instanceof Error ? error.message : 'Unknown error',
    };

    res.status(500).json(response);
  }
});

/**
 * PUT /api/v1/rent-agreements/:id
 * Update rent agreement
 */
router.put('/:id', requireActiveUser, validate(schemas.rentAgreementUpdate), async (req: Request, res: Response) => {
  try {
    const agreementId = parseInt(req.params.id);
    
    if (isNaN(agreementId)) {
      const response: ApiResponse = {
        success: false,
        message: 'Invalid agreement ID',
      };
      return res.status(400).json(response);
    }

    const currentUser = getCurrentUser(req);
    
    if (!currentUser) {
      const response: ApiResponse = {
        success: false,
        message: 'User not found',
      };
      return res.status(404).json(response);
    }

    // Check if agreement exists and user is authorized
    const existingAgreement = await db.rentAgreement.findById(agreementId);
    
    if (!existingAgreement) {
      const response: ApiResponse = {
        success: false,
        message: 'Rent agreement not found',
      };
      return res.status(404).json(response);
    }

    if (existingAgreement.tenantId !== currentUser.id) {
      const response: ApiResponse = {
        success: false,
        message: 'Unauthorized to update this agreement',
      };
      return res.status(403).json(response);
    }

    const updatedAgreement = await db.rentAgreement.update(agreementId, req.body);
    
    if (!updatedAgreement) {
      const response: ApiResponse = {
        success: false,
        message: 'Failed to update rent agreement',
      };
      return res.status(500).json(response);
    }

    const response: ApiResponse<RentAgreement> = {
      success: true,
      data: updatedAgreement,
      message: 'Rent agreement updated successfully',
    };

    res.status(200).json(response);
  } catch (error) {
    logger.error('Update rent agreement failed:', error);
    
    const response: ApiResponse = {
      success: false,
      message: 'Failed to update rent agreement',
      error: error instanceof Error ? error.message : 'Unknown error',
    };

    res.status(500).json(response);
  }
});

/**
 * DELETE /api/v1/rent-agreements/:id
 * Delete rent agreement
 */
router.delete('/:id', requireActiveUser, async (req: Request, res: Response) => {
  try {
    const agreementId = parseInt(req.params.id);
    
    if (isNaN(agreementId)) {
      const response: ApiResponse = {
        success: false,
        message: 'Invalid agreement ID',
      };
      return res.status(400).json(response);
    }

    const currentUser = getCurrentUser(req);
    
    if (!currentUser) {
      const response: ApiResponse = {
        success: false,
        message: 'User not found',
      };
      return res.status(404).json(response);
    }

    // Check if agreement exists and user is authorized
    const existingAgreement = await db.rentAgreement.findById(agreementId);
    
    if (!existingAgreement) {
      const response: ApiResponse = {
        success: false,
        message: 'Rent agreement not found',
      };
      return res.status(404).json(response);
    }

    if (existingAgreement.tenantId !== currentUser.id) {
      const response: ApiResponse = {
        success: false,
        message: 'Unauthorized to delete this agreement',
      };
      return res.status(403).json(response);
    }

    const deleted = await db.rentAgreement.delete(agreementId);
    
    if (!deleted) {
      const response: ApiResponse = {
        success: false,
        message: 'Failed to delete rent agreement',
      };
      return res.status(500).json(response);
    }

    const response: ApiResponse = {
      success: true,
      message: 'Rent agreement deleted successfully',
    };

    res.status(200).json(response);
  } catch (error) {
    logger.error('Delete rent agreement failed:', error);
    
    const response: ApiResponse = {
      success: false,
      message: 'Failed to delete rent agreement',
      error: error instanceof Error ? error.message : 'Unknown error',
    };

    res.status(500).json(response);
  }
});

/**
 * GET /api/v1/rent-agreements/property/:propertyId
 * Get rent agreements for a specific property
 */
router.get('/property/:propertyId', requireActiveUser, async (req: Request, res: Response) => {
  try {
    const propertyId = parseInt(req.params.propertyId);
    
    if (isNaN(propertyId)) {
      const response: ApiResponse = {
        success: false,
        message: 'Invalid property ID',
      };
      return res.status(400).json(response);
    }

    const agreements = await db.rentAgreement.findByPropertyId(propertyId);

    const response: ApiResponse<RentAgreement[]> = {
      success: true,
      data: agreements,
    };

    res.status(200).json(response);
  } catch (error) {
    logger.error('Get property rent agreements failed:', error);
    
    const response: ApiResponse = {
      success: false,
      message: 'Failed to get property rent agreements',
      error: error instanceof Error ? error.message : 'Unknown error',
    };

    res.status(500).json(response);
  }
});

export default router;
